ui/src/app/streams/flo/editor.service.ts

```diff
@@ -795,87 +795,86 @@ export class EditorService implements Flo.Editor {
-                           linksIn = graph.getConnectedLinks(element, {inbound: true});
-                           if (linksIn.length === 0) {
-                               openPorts.push({'node': element, 'openPort': 'input'});
-                           }
-                       }
-                       break;
-                   default:
-               }
-            });
+            this.performAutoLinking(flo, dragDescriptor.source.view);
+        }
+    }
 
-            if (openPorts.length === 1) {
-                // One candidate match!
-                const match = openPorts.shift();
-                // TODO: replace selector CSS class with the result of view.getSelector(...)
-                if (match.openPort === 'input') {
-                    flo.createLink({
-                        'id': source.id,
-                        'selector': '.output-port', // /.output-port/.input-port
-                        'port': 'output' // input/output
-                    }, {
-                        'id': match.node.id,
-                        'selector': '.input-port',
-                        'port': 'input'
-                    });
-                } else { // match.openPort === 'output'
-                    flo.createLink({
-                        'id': match.node.id,
-                        'selector': '.output-port', // /.output-port/.input-port
-                        'port': 'output' // input/output
-                    }, {
-                        'id': source.id,
-                        'selector': '.input-port',
-                        'port': 'input'
-                    });
-                }
+    private performAutoLinking(flo: Flo.EditorContext, view: dia.CellView) {
+      // Search the graph for open ports - if only 1 is found, and it matches what
+      // the dropped node needs, join it up!
+      const group = view.model.attr('metadata/group');
+      const graph = flo.getGraph();
+      let wantOpenInputPort = false; // want something with an open input port
+      let wantOpenOutputPort = false; // want something with an open output port
+      if (group === ApplicationType[ApplicationType.source]) {
+        wantOpenInputPort = true;
+      } else if (group === ApplicationType[ApplicationType.sink]) {
+        wantOpenOutputPort = true;
+      } else if (group === ApplicationType[ApplicationType.processor]) {
+        wantOpenInputPort = true;
+        wantOpenOutputPort = true;
+      }
+      const openPorts = [];
+      let linksIn, linksOut;
+
+      graph.getElements().filter(e => e.id !== view.model.id && e.attr('metadata/name')).forEach(element => {
+        switch (element.attr('metadata/group')) {
+          case ApplicationType[ApplicationType.source]:
+            if (wantOpenOutputPort) {
+              const primaryLink = graph.getConnectedLinks(element, {outbound: true}).find(l => !l.attr('props/isTapLink'));
+              linksOut = this.getOutgoingStreamLinks(graph, element);
```

Comment:
Not sure linksOut is needed, seems to be a leftover that was replaced by the primaryLink code before it. linksOut is set but never read.